internal/cluster/distribution/eks/eksprovider/workflow/select_volume_size_activity.go

```diff
@@ -0,0 +1,70 @@
+// Copyright © 2020 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workflow
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"emperror.dev/errors"
+)
+
+const (
+	SelectVolumeSizeActivityName = "eks-select-volume-size-activity"
+)
+
+type SelectVolumeSizeActivity struct {
+	defaultVolumeSize int
+	unsetVolumeSize   int
+}
+
+type SelectVolumeSizeActivityInput struct {
+	EKSActivityInput
+	AMIImageSize int
+}
+
+type SelectVolumeSizeActivityOutput struct {
+	VolumeSize int
+}
+
+func NewSelectVolumeSizeActivity(defaultVolumeSize, unsetVolumeSize int) (activity *SelectVolumeSizeActivity) {
+	return &SelectVolumeSizeActivity{
+		defaultVolumeSize: defaultVolumeSize,
+		unsetVolumeSize:   unsetVolumeSize,
+	}
+}
+
+func (activity *SelectVolumeSizeActivity) Execute(ctx context.Context, input SelectVolumeSizeActivityInput) (output *SelectVolumeSizeActivityOutput, err error) {
```

Comment:
Looking at this code: it does not necessarily have to be an activity: it's deterministic. It could just be a function called by the workflow. That said, we can leave it in an activity for now, it doesn't hurt. Up to you.

Reply:
The reason I put it in an activity is to be able to initialize it with the default value (and optionally the unset value) at registration time (top level as a system preconfiguration), because the default node volume size is not available at the workflow level without injecting it there at registration time as well and passing the value to the function - same with a non-activity object wrapping the selection logic and being stored in the workflow.

I'm a little bit afraid of the scalability of involving workflows in such a static logical configurations of subcomponents, it doesn't sound scalable to me in the long run, I like isolating the lower level business logic from the main responsibilities of the workflow which in my mind are
1. activity aggregatiion/integration ,
2. activity output/input data transformation,
3. high level business logic conditionality branching (optional or input-based activity groups)

Basically what I expect to see when I look at a workflow is the following
-> workflowInput ->
-> firstActivityOutput = FirstActivity(workflowInput.Parts) ->
-> secondActivityOutput = SecondActivity(transform(workflowInput.OtherParts, firstOutput)) ->
-> if condition([workflowInput.DifferentParts][, firstActivityOutput][, secondActivityOutput])
    then 
        thirdActivityOutput = ThirdActivity(transform([workflowInput.SomeParts][, firstActivityOutput][, secondActivityOutput])) ->
    else
        fourthActivityOutput = FourthActivity(transform([workflowInput.SomeParts][, firstActivityOutput][, secondActivityOutput])) ->
...
Also activities could be substituted with child workflows in these cases if there is a logical grouping of activities which is worth moving to workflows for reusability.
And in the pseudo-code above `transform` in my mind is supposed to be structural reorganization only (maps to slices, vice versa, dropping fields or aggregating information of different origin) and no business logic.

(This is not necessarily the current state in the Pipeline, but that's the best "desired case" for me regarding workflows, whether it is realistic is up for debate.)

I'm also a little bit afraid that the activity overhead is not worth it for such simple logics as in this use-case as the selection currently is, but I didn't want to prematurely be anxious about this.

Reply:
As I said, it doesn't make much of a difference. Workflow is responsible for coordinating the business logic, that's why they are/should be heavily unit tested, so a condition like this fits perfectly into the workflow code.

Think of it this way: you could substitute each activity call with a plain function call that hides activities. Then your workflow would simply be a function with business logic in it.

Let's leave it in an activity for now.